packages/byods/src/data-source-service/index.ts

```diff
@@ -0,0 +1,133 @@
+/* eslint-disable class-methods-use-this */
+/* eslint-disable import/no-extraneous-dependencies */
+
+import {v4 as uuidv4} from 'uuid';
+import {decodeJwt, JWTPayload} from 'jose';
+import DataSourceClient from '../data-source-client/index';
+
+export default class DataSourceService {
+  private dataSourceClient: DataSourceClient;
+  private timer: any;
+
+  /**
+   * Creates an instance of DataSourceService.
+   * @param {DataSourceClient} dataSourceClient - The  dataSourceClient instance to use for DataSourceService.
+   * @example
+   * const dataSourceService = new DataSourceService(dataSourceClient);
+   */
+
+  constructor(dataSourceClient: DataSourceClient) {
+    this.dataSourceClient = dataSourceClient;
+  }
+
+  /**
+   * This method refreshes the DataSource token using dataSourceId and tokenLifetimeMinute
+   * @param {string} dataSourceId The id of data source
+   * @param {number} tokenLifetimeMinutes The Life time minutes for the data source. <=1440
+   * @returns {Promise<void>}
+   */
+
+  public async refreshDataSourceToken(
+    dataSourceId: string,
+    tokenLifetimeMinutes?: number
+  ): Promise<void> {
+    try {
+      if (dataSourceId) {
+        if (tokenLifetimeMinutes) {
+          this.startAutoRefresh(dataSourceId, tokenLifetimeMinutes - 5);
+        } else {
+          console.log('tokenLifetimeMinutes is not provided, setting up default value');
+          tokenLifetimeMinutes = 55;
+          this.startAutoRefresh(dataSourceId, tokenLifetimeMinutes);
+        }
+      } else {
+        console.log('Required field is missing, Please pass dataSourceId.');
+      }
+    } catch (error) {
+      console.log('Encountered some error, error is', error);
+    }
+  }
+
+  /**
+   * This Private method will start auto refreshing the DataSource token with interval as tokenLifetimeMinutes.
+   * @param {string} dataSourceId The id of data source
+   * @param {number} tokenLifetimeMinutes The Life time minutes for the data source. <=1440
+   * @returns {Promise<void>}
+   */
+
+  private async startAutoRefresh(
+    dataSourceId: string,
+    tokenLifetimeMinutes?: number
+  ): Promise<void> {
+    try {
+      // Call the get method to fetch the dataSource details & then call the update method to update the token!
+      const payloadForDataSourceUpdateMethod: any = {};
+      await this.dataSourceClient
+        .get(dataSourceId)
+        .then((response: any) => {
+          const jwsToken: any = response.jwsToken;
+          const jwsTokenPayload: any = this.decodeJWSTokenAndGetPayload(jwsToken);
+          payloadForDataSourceUpdateMethod.schemaId = response.schemaId;
+          payloadForDataSourceUpdateMethod.url = jwsTokenPayload.payload.com.cisco.datasource.url;
+          payloadForDataSourceUpdateMethod.subject = jwsTokenPayload.payload.sub;
+          payloadForDataSourceUpdateMethod.audience = jwsTokenPayload.payload.aud;
+          payloadForDataSourceUpdateMethod.nonce = uuidv4();
```

Comment:
This seems to always use uuidv4() for nonce generation. We should also allow the developer to provide their own nonce generator. We should default to uuidv4 only if a nonce generator is not provided

Reply:
**_Addressed!_**